objc/include/onnxruntime/ort_env.h

```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import <Foundation/Foundation.h>
+
+/**
+ * The ORT environment.
+ */
+@interface ORTEnv : NSObject
+
+/**
+ * Creates an ORT Environment.
+ *
+ * @param error Optional error information set if an error occurs.
+ * @return The instance, or nil if an error occurs.
+ */
+- (instancetype)initWithError:(NSError**)error;
+
```

Comment:
could this just be called 'init'? not sure we'd ever have a version where there's no error reporting. 

general question: are we modelling this on the C API or C++ API? if the C++ API, should an init that fails throw instead?

Reply:
This kind of init naming is Apple convention, 
[https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html)

Reply:
we should probably expose a conventional Objective-C API (but I'm still learning what that is).

regarding exceptions, this is a quote from [here](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/ErrorHandling/ErrorHandling.html):
When you’re writing code with Objective-C, exceptions are used solely for programmer errors, like out-of-bounds array access or invalid method arguments. These are the problems that you should find and fix during testing before you ship your app.


---
In reply to: [616189636](https://github.com/microsoft/onnxruntime/pull/7366#discussion_r616189636) [](ancestors = 616189636)

Reply:
The NSError is somewhat defacto though. I can maybe understand something like initWithThreadingOptions or initWithLogLevel. 

That said, all the Apple classes I randomly looked at don't follow that approach for the 'init' methods, so for that specific use case it seems to be ok to just over various overrides called 'init'. e.g. [https://developer.apple.com/documentation/mlcompute/mlctensor](https://developer.apple.com/documentation/mlcompute/mlctensor) or [https://developer.apple.com/documentation/appkit/nswindow](https://developer.apple.com/documentation/appkit/nswindow)

---
In reply to: [616207853](https://github.com/microsoft/onnxruntime/pull/7366#discussion_r616207853) [](ancestors = 616207853)

Reply:
Sorry - just noticed that it was defaulting to the Swift API. in Objective-C the equivalent of the 'init' overrides is a static create method. 

---
In reply to: [616262049](https://github.com/microsoft/onnxruntime/pull/7366#discussion_r616262049) [](ancestors = 616262049,616207853)

Reply:
looked around for some naming conventions. Apple has some:
[https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html)
"Make the word before the argument describe the argument."

actually this is the same link that @gwang-msft mentioned above.

so between `init:` and `initWithError:`, I think the latter is preferred. will leave it as it is for now.

Reply:
also found out there is a way to disable the `init` inherited from NSObject (using `NS_UNAVAILABLE`). will do that